ios/ReactNativeShareExtension.m

```diff
@@ -91,10 +92,31 @@ - (void)extractDataFromContext:(NSExtensionContext *)context withCallback:(void(
             }];
         } else if (imageProvider) {
             [imageProvider loadItemForTypeIdentifier:IMAGE_IDENTIFIER options:nil completionHandler:^(id<NSSecureCoding> item, NSError *error) {
-                NSURL *url = (NSURL *)item;
-
+                
+                /**
+                 * Save the image to NSTemporaryDirectory(), which cleans itself tri-daily.
+                 * This is necessary as the iOS 11 screenshot editor gives us a UIImage, while
+                 * sharing from Photos and similar apps gives us a URL
+                 * Therefore the solution is to save a UIImage, either way, and return the local path to that temp UIImage
+                 * This path will be sent to React Native and can be processed and accessed RN side.
+                **/
+                
+                UIImage *sharedImage;
+                NSString *filePath = [NSTemporaryDirectory() stringByAppendingPathComponent:@"RNSE_TEMP_IMG"];
+                NSString *fullPath = [filePath stringByAppendingPathExtension:@"png"];
+                
+                if ([(NSObject *)item isKindOfClass:[UIImage class]]){
+                    sharedImage = (UIImage *)item;
+                }else if ([(NSObject *)item isKindOfClass:[NSURL class]]){
+                    NSURL* url = (NSURL *)item;
+                    NSData *data = [NSData dataWithContentsOfURL:url];
+                    sharedImage = [UIImage imageWithData:data];
+                }
+                
+                [UIImagePNGRepresentation(sharedImage) writeToFile:fullPath atomically:YES];
```

Comment:
Share extension might crash due to memory pressure with UIImageXXXRepresentation. On an iPad Air, I have an image taken from my iPhone that kills the SE every time it hits this line. I'm testing out using ImageIO, will try to remember to report back if use that API helps.

Reply:
Using CGImageDestinationCreateWithURL(), there's at least output explicitly stating the crash is memory related. A save from NSData seems to work. Can't think of an issue with doing it this way, but still kind of bouncing this issue around in me brain.

```
NSError *error;
[data writeToFile:fullPath options:NSDataWritingAtomic error:&error];

NSFileManager *fileManager = [NSFileManager defaultManager];
if ([fileManager fileExistsAtPath:fullPath]){
  NSLog(@"NSData writeToFile successful");
} else {
  NSLog(@"NSData writeToFile failed");
}
```

Reply:
Using NSData instead of UIImage works well in my local testing. I'll testing with the screenshot editor, a bit later today.

Reply:
Good find - I was skeptical about this methods memory usage when I first wrote it..

Make sure to report back with the testing, I'll be interested to know if it works smoothly on all devices.

Reply:
NSData is the preferred route. Here's a more complete fix, sans exception handling (I'm handling exceptions differently than this library is likely to):

```
if ([(NSObject *)item isKindOfClass:[UIImage class]]){
    UIImage *sharedImage = (UIImage *)item;

    NSString *imageType = @"png";
           
    NSString *fileName = [[NSUUID UUID] UUIDString];
    NSString *filePath = [NSTemporaryDirectory() stringByAppendingPathComponent:fileName];
    NSString *fullPath = [filePath stringByAppendingPathExtension:imageType];

    NSData *imageData = UIImagePNGRepresentation(sharedImage);

    [imageData writeToFile:fullPath atomically:YES];

    if ([fileManager fileExistsAtPath:fullPath]) {
      NSLog(@"NSData writeToFile successful");
    } else {
      RCTLog(@"ERROR (extractDataFromContext): NSData writeToFile failed for tmp UIImage item");
//      TODO: Throw: file doesn't exist.
    }
}else if ([(NSObject *)item isKindOfClass:[NSURL class]]){
    NSURL* url = (NSURL *)item;

    NSString *imageType = [NSMutableString stringWithString:[[[url absoluteString] pathExtension] lowercaseString]];

    NSString *fileName = [[url lastPathComponent] stringByDeletingPathExtension];
    NSString *filePath = [NSTemporaryDirectory() stringByAppendingPathComponent:fileName];
    NSString *fullPath = [filePath stringByAppendingPathExtension:imageType];
    
    NSData *data = [NSData dataWithContentsOfURL:url];
    [data writeToFile:fullPath atomically:YES];

    if ([fileManager fileExistsAtPath:fullPath]) {
      NSLog(@"NSData writeToFile successful for tmp NSURL");
    } else {
      RCTLog(@"ERROR (extractDataFromContext): NSData writeToFile failed for NSURL item");
//      TODO: Throw: file doesn't exist.
    }
}

if(callback) {
  callback(fullPath, [fullPath pathExtension], nil);
}
```

Reply:
@bluey31 All tests passed with iPhone 7 and iPad Air on iOS 10.2 and 11; works fine in a TF release. Let me know if you run into any issues or need any help.